### PR TITLE
Add support test262 $262 global object support

### PR DESCRIPTION
**Pull request by kisbg**
Few test-cases in test262 uses a $262 object to run a few method (for example detachedArrayBuffer)

JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: .github/workflows/gh-actions.yml [hunk 1]
```diff
@@ -90,6 +90,8 @@ jobs:
 
   Conformance_Tests_ESNext_A:
     runs-on: ubuntu-20.04
+    env:
+      BUILD_OPTIONS: --buildoptions=--test262=ON
```
Comment (ossy-szeged) [inline .github/workflows/gh-actions.yml:94]: run-tests --test262-esnext should implicitly trigger building with --test262=ON. We should be able to run test262 tests locally without any additional command line option.

Comment (kisbg) [reply .github/workflows/gh-actions.yml:94]: i see, thank you for the information !

File: docs/01.CONFIGURATION.md [hunk 2]
```diff
@@ -255,6 +255,17 @@ This option is disabled by default.
...
 | Python: | `--system-allocator=ON/OFF`                  |
 
+### Test262
+
+This option enables the use of $262 object.
+This option is disabled by default.
+
+| Options |                                              |
+|---------|----------------------------------------------|
+| C:      | `-JERRY_TEST262=0/1`               |
```
Comment (galpeter) [inline docs/01.CONFIGURATION.md:265]: * The closing `|` should be aligned in all lines.
* Missing a `D` prefix: `-DJERRY_TEST262=...`

File: jerry-core/CMakeLists.txt [hunk 3]
```diff
@@ -36,6 +36,7 @@ set(JERRY_REGEXP_DUMP_BYTE_CODE     OFF          CACHE BOOL   "Enable regexp byt
 set(JERRY_SNAPSHOT_EXEC             OFF          CACHE BOOL   "Enable executing snapshot files?")
 set(JERRY_SNAPSHOT_SAVE             OFF          CACHE BOOL   "Enable saving snapshot files?")
 set(JERRY_SYSTEM_ALLOCATOR          OFF          CACHE BOOL   "Enable system allocator?")
+set(JERRY_TEST262                   OFF          CACHE BOOL   "Enable test262?")
```
Comment (galpeter) [inline jerry-core/CMakeLists.txt:39]: Maybe a better message would be `Enable test262 support?`

File: jerry-main/main-utils.c [hunk 4]
```diff
@@ -44,6 +44,183 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+static void
+register_test262 (void)
+{
+  jerry_value_t global_obj = jerry_get_global_object ();
+  jerry_value_t test262_object = jerry_create_object ();
+
+  test262_register_function (test262_object, "createRealm", test262_create_realm);
+  test262_register_function (test262_object, "detachArrayBuffer", test262_detach_array_buffer);
+  test262_register_function (test262_object, "evalScript", test262_eval_script);
+  test262_register_function (test262_object, "gc", test262_gc);
```
Comment (galpeter) [inline jerry-main/main-utils.c:205]: For the gc we already have a handler `jerryx_handler_gc`

File: jerry-main/main-utils.c [hunk 5]
```diff
@@ -44,6 +44,183 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+  if (args_cnt < 1 || !jerry_value_is_string (args_p[0]))
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *) "Expected a string");
+  }
+
+  jerry_size_t str_size = jerry_get_utf8_string_size (args_p[0]);
+  jerry_char_t *str_buf_p = jerry_heap_alloc (str_size * sizeof (jerry_char_t));
+
+  if (str_buf_p == NULL
+     || jerry_string_to_utf8_char_buffer (args_p[0], str_buf_p, str_size) != str_size)
```
Comment (galpeter) [inline jerry-main/main-utils.c:148]: incorrect indentation.

File: jerry-main/main-utils.c [hunk 6]
```diff
@@ -44,6 +44,183 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+  (void) func_obj_val; /* unused */
+  (void) this_p; /* unused */
+
+  if (args_cnt < 1 || !jerry_value_is_string (args_p[0]))
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *) "Expected a string");
+  }
+
+  jerry_size_t str_size = jerry_get_utf8_string_size (args_p[0]);
+  jerry_char_t *str_buf_p = jerry_heap_alloc (str_size * sizeof (jerry_char_t));
```
Comment (galpeter) [inline jerry-main/main-utils.c:145]: I'm not sure it is a good idea to use heap alloc here. What are we using when loading normal scripts form files?

File: jerry-main/main-utils.c [hunk 7]
```diff
@@ -44,6 +44,183 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+
+  test262_register_function (test262_object, "createRealm", test262_create_realm);
+  test262_register_function (test262_object, "detachArrayBuffer", test262_detach_array_buffer);
+  test262_register_function (test262_object, "evalScript", test262_eval_script);
+  test262_register_function (test262_object, "gc", test262_gc);
+
+  jerry_value_t prop_name = jerry_create_string ((const jerry_char_t *) "global");
+  jerry_set_property (test262_object, prop_name, global_obj);
+
+  // register_report_error (result_val, "global");
```
Comment (galpeter) [inline jerry-main/main-utils.c:210]: are these comments required?

**Comment by kisbg** [comment]
@galpeter  Thank your for the review! I updated the pull request based on your review.

File: .github/workflows/gh-actions.yml [hunk 8]
```diff
@@ -144,4 +144,3 @@ jobs:
         run: sudo apt-get install gcc-multilib
       - name: UBSAN - jerry-tests
         run: $RUNNER --jerry-tests $SKIP_LIST $BUILD_OPTIONS -q
-
```
Comment (rerobika) [inline .github/workflows/gh-actions.yml:147]: Please do not remove this line.

File: jerry-main/main-utils.c [hunk 9]
```diff
@@ -44,6 +44,126 @@ main_register_global_function (const char *name_p, /**< name of the function */
   jerry_release_value (result_val);
 } /* main_register_global_function */
 
+#if defined (JERRY_TEST262) && (JERRY_TEST262 == 1)
+
+static void
+test262_register_function (jerry_value_t test262_obj, /** $262 object */
```
Comment (rerobika) [inline jerry-main/main-utils.c:50]: Missing function description.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-main/main-utils.c [hunk 10]
```diff
@@ -44,6 +44,129 @@ main_register_global_function (const char *name_p, /**< name of the function */
   jerry_release_value (result_val);
 } /* main_register_global_function */
 
+#if defined (JERRY_TEST262) && (JERRY_TEST262 == 1)
+
+/**
+ * Register a test262 function in the $262 object.
```
Comment (rerobika) [inline jerry-main/main-utils.c:50]: Register a method for the $262 object.

**Review by rerobika — CHANGES_REQUESTED** [review]

**Review by rerobika — APPROVED** [review]
LGTM

**Comment by rerobika** [comment]
Please rebase the PR.

File: jerry-main/main-utils.c [hunk 11]
```diff
@@ -44,6 +44,129 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *) "Expected a string");
+  }
+
+  jerry_size_t str_size = jerry_get_utf8_string_size (args_p[0]);
+  jerry_char_t *str_buf_p = malloc (str_size * sizeof (jerry_char_t));
+
+  if (str_buf_p == NULL || jerry_string_to_utf8_char_buffer (args_p[0], str_buf_p, str_size) != str_size)
+  {
+    return jerry_create_error (JERRY_ERROR_RANGE, (jerry_char_t *) "Internal error");
```
Comment (galpeter) [inline jerry-main/main-utils.c:122]: If the malloc was successful here we need to call free on it.

File: jerry-main/main-utils.c [hunk 12]
```diff
@@ -44,6 +44,129 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+
+  test262_register_function (test262_object, "detachArrayBuffer", test262_detach_array_buffer);
+  test262_register_function (test262_object, "evalScript", test262_eval_script);
+  test262_register_function (test262_object, "gc", jerryx_handler_gc);
+
+  jerry_value_t prop_name = jerry_create_string ((const jerry_char_t *) "global");
+  jerry_value_t result = jerry_set_property (test262_object, prop_name, global_obj);
+
+  jerry_release_value (prop_name);
+  assert (!jerry_value_is_error (result));
```
Comment (galpeter) [inline jerry-main/main-utils.c:156]: swap the lines:
```c
assert(..)
.._release_..(...)
```
and there is no need for an empty line after the result.

File: jerry-main/main-utils.c [hunk 13]
```diff
@@ -44,6 +44,129 @@ main_register_global_function (const char *name_p, /**< name of the function */
...
+
+  prop_name = jerry_create_string ((const jerry_char_t *) "$262");
+  result = jerry_set_property (global_obj, prop_name, test262_object);
+
+  assert (!jerry_value_is_error (result));
+  jerry_release_value (prop_name);
```
Comment (galpeter) [inline jerry-main/main-utils.c:162]: same as above.

File: jerry-main/main-utils.c (hunk 12 above)
Comment (galpeter) [reply jerry-main/main-utils.c:156]: and afaik the result should be released before re-using.

File: jerry-main/main-utils.c (hunk 13 above)
Comment (galpeter) [reply jerry-main/main-utils.c:162]: result should be freed directly after (not as the last element in the method.

**Comment by ossy-szeged** [comment]
Please close already resolved reviews with "Resolve conversation" button.

File: jerry-main/main-utils.c [hunk 14]
```diff
@@ -44,6 +44,129 @@ main_register_global_function (const char *name_p, /**< name of the function */
   jerry_release_value (result_val);
 } /* main_register_global_function */
 
+#if defined (JERRY_TEST262) && (JERRY_TEST262 == 1)
```
Comment (zherczeg) [inline jerry-main/main-utils.c:47]: What about doing this without a define? This is a simple test program, I don't see any issue to always have this comannd line feature.

Comment (zherczeg) [reply jerry-main/main-utils.c:47]: This question still stands.

File: jerry-main/CMakeLists.txt [hunk 15]
```diff
@@ -56,6 +56,7 @@ macro(jerry_create_executable JERRY_NAME)
   install(TARGETS ${JERRY_NAME} DESTINATION bin)
 endmacro()
 
+
```
Comment (rerobika) [inline jerry-main/CMakeLists.txt:59]: Please remove this new line.

**Review by rerobika — APPROVED** [review]
LGTM after this minor change

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**